### PR TITLE
*: support switch auth when change user

### LINKED ISSUE #38829: ChangeUser does not support switch auth
## Bug Report

ref pingcap/tiproxy#127

### 4. What is your TiDB version? (Required)

caa26a4

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #38829

Problem Summary: Support switch auth request when changing user. This allows mismatch between client/server auth methods. And it gives a chance for proxy to change the correct salt.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```
tiup --db.binpath ./tidb-server
./bin/tiproxy
mvn "-Dcom.mysql.jdbc.testsuite.url=jdbc:mysql://127.0.0.1:6000/test?user=root&password=&useSSL=false&useServerPrepStmts=true" '-Dtest=ConnectionRegressionTest#testBug19354014' test
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
support switch auth request for change user command
```

### CODE REVIEW

File: server/conn.go [hunk 1]
```diff
@@ -2458,8 +2458,19 @@ func (cc *clientConn) handleChangeUser(ctx context.Context, data []byte) error {
...
 	cc.dbname = string(hack.String(dbName))
+	pluginName := ""
+	if len(data) > 0 {
+		// skip character set
+		if cc.capability&mysql.ClientProtocol41 > 0 && len(data) >= 2 {
+			data = data[2:]
+		}
+		if cc.capability&mysql.ClientPluginAuth > 0 && len(data) > 0 {
+			pluginNameB, _ := parseNullTermString(data)
+			pluginName = string(pluginNameB)
```
Comment (djshow832) [inline server/conn.go:2471]:
```suggestion
			pluginName = string(hack.String(pluginNameB))
```
Avoid copying strings like the other usages in this function.

**Review by djshow832 — COMMENTED** [review]
Can you add a UT in case someone breaks it in the future?

**Review by CbcWestwolf — APPROVED** [review]

**Review by dveeden — APPROVED** [review]
Looks to match with https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_com_change_user_auth

Not all CI checks passed.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**